Add FeeAccount to StdTx and add fee-delegation module

aaronc opened the pull request:
Adds:
* The `FeeAccount sdk.AccAddress` field to `StdTx`
* The required logic in `AnteHandler` to deduct fees from the `FeeAccount` if one is provided and that account has indeed delegated fees to the account signing the transaction
* A `fee-delegation` module to manage granting of fee allowances. `FeeAllowance` is an interface which allows for different types of fee allowance policies such as daily or hourly limits
* CLI support for a `--fee-account` flag

Use cases:
- an organization can delegate fees to employees from the master organizational account so that employees can sign certain transactions onto the blockchain without needing to worry about wallets
- a user could keep a less trusted key on their phone and delegate fees to this key for purposes such as voting on proposals (this would also use generic action delegation which was worked on in the hackathon and will be submitted as a separate PR)
- a third party service could delegate fees to users who pay off-chain via fiat or other crypto so that those users can use Dapps without needing to fill their wallets

This PR is squashes many commits from the Gaian's team at Hackatom Berlin. It is now being worked on in the context of the Community Group on Key Management (https://github.com/cosmos-cg-key-management) and Regen Network intends to test it in on its live testnet in the coming weeks. Once the testing is complete and the community has had a time to discuss the design more thoroughly, we intend to move this PR out of draft status. In the meantime, this thread can be a forum for discussion and review of the proposed changes.

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry: `clog add [section] [stanza] [message]`
- [ ] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

AdityaSripal commented on x/fee-delegation/keeper.go:
```diff
@@ -0,0 +1,81 @@
+	if allowance == nil {
+		return false
+	}
+	allow, updated, delete := allowance.Accept(fee, ctx.BlockHeader())
```
Must set `updated` back into store with the same key or the changes do not get persisted to application store.

aaronc replied:
That happens below on line 78 below actually. I'll add a few code comments to make that clear

AdityaSripal commented on x/auth/ante.go:
```diff
@@ -122,18 +142,21 @@ func NewAnteHandler(ak AccountKeeper, supplyKeeper types.SupplyKeeper, sigGasCon
 		}
 
+		if len(stdTx.FeeAccount) != 0 {
+			if err := feeAccount.SetSequence(feeAccount.GetSequence() + 1); err != nil {
```
Don't think its necessary to update `Sequence` of FeeAccount. It didn't sign anything

aaronc replied:
Right!

AdityaSripal commented:
Thanks for working on this! 

Is there a reason you decided to go with this approach of allowing Fee Grantees to deduct directly from the Fee Granter account as opposed to the following:

1. Granter allows FeeDelegation of 50 atoms
2. 50 atoms gets deducted from his/her Account and get escrowed into an isolated ModuleAccount or stored in FeeDelegationKeeper
3. When Grantee asks to spend fees from Granter Account, the escrowed account is checked for corresponding funds and has the appropriate amount deducted
4. RevokeDelegation returns balance of escrowed funds to Granter

aaronc commented:
> Thanks for working on this!
> 
> Is there a reason you decided to go with this approach of allowing Fee Grantees to deduct directly from the Fee Granter account as opposed to the following:
> 
> 1. Granter allows FeeDelegation of 50 atoms
> 2. 50 atoms gets deducted from his/her Account and get escrowed into an isolated ModuleAccount or stored in FeeDelegationKeeper
> 3. When Grantee asks to spend fees from Granter Account, the escrowed account is checked for corresponding funds and has the appropriate amount deducted
> 4. RevokeDelegation returns balance of escrowed funds to Granter

The disadvantage of transferring into an escrow account is that this needs to be done for every granter/grantee pair and that may not be economically optimal for a granter who is allocating say daily spend limits to a lot of grantees. A common scenario that I am imagining is that an organization wants to create a fee pool for their employees or a third party service wants to offer this for a fee (in some other currency), and that these grants would be on a monthly/daily/hourly spend limit basis. So a granter might want to over-extend their total max daily allocation to a pool given actual usage and only recharge as needed. If I give 100 people a 10 atom daily spend limit for a validity of say 30 days, I shouldn't need to give each user 300 atoms in escrow (a total of 30000 atoms). I can just have one account that has say 1000 atoms max at a time, monitor usage and fill it up as needed. Those 100 people may only spend 5000 atoms total over the course of a month, but if I did it as escrow upfront I might need to lockup 30000 atoms. So in my reasoning, letting people pool things into one account they fill up as needed allows for a more economically practical management of resources.

alexanderbez commented:
@aaronc thanks for the PR! I'd be happy to review this and provide feedback, but do you mind first updating the PR to reflect the latest changes on `master`?

aaronc commented:
> @aaronc thanks for the PR! I'd be happy to review this and provide feedback, but do you mind first updating the PR to reflect the latest changes on `master`?

Thanks @alexanderbez. Just rebased again.

In my mind the most important thing to review currently are the changes to `ante.go` and `stdtx.go`. Other details about fee delegation are important, but can of course be swapped in and out with different modules.

Also, I forgot to mention in my previous comment the sub-key use case. In this scenario a user would have a single account and would delegate fees to other "sub keys". So it's one account, just different keys. Other features to support sub-keys will be coming in 2 PR's that I am preparing. This has been discussed on telegram with others including B-harvest, and would likely supersede the previous sub-key proposal being more general to cover other cases like the organizational use case described above.

alexanderbez commented on x/fee-delegation/allowances.go:
```diff
@@ -0,0 +1,80 @@
+	PeriodReset time.Time
+}
+
+func (allowance PeriodicFeeAllowance) Accept(fee sdk.Coins, block abci.Header) (allow bool, updated FeeAllowance, remove bool) {
```
ditto

alexanderbez commented on x/fee-delegation/allowances.go:
```diff
@@ -0,0 +1,80 @@
+	}
+	periodReset := allowance.PeriodReset
+	periodCanSpend := allowance.PeriodCanSpend
+	if periodReset.Before(block.Time) {
```
lets add a comment to what we're checking and doing here.

alexanderbez commented on x/fee-delegation/cli_query.go:
```diff
@@ -0,0 +1,46 @@
+				return err
+			}
+
+			fmt.Println(string(res))
```
Let's use `return cliCtx.PrintOutput(...)` instead here. That means we'll have to decode `res` into the correct type.

alexanderbez commented on x/fee-delegation/cli_tx.go:
```diff
@@ -0,0 +1,64 @@
+func GetCmdDelegateFees(cdc *codec.Codec) *cobra.Command {
+	return &cobra.Command{
+		Use:   "delegate [grantee] [fee-allowance]",
+		Short: "delegate",
```
should the nomenclature perhaps be `grant` instead?

alexanderbez commented on x/fee-delegation/genesis.go:
```diff
@@ -0,0 +1,35 @@
+
+// GenesisState defines genesis data for the module
+type GenesisState struct {
+	// Delegations []Delegation `json:"delegations"`
```
??? Is this still to be implemented?

alexanderbez commented on x/fee-delegation/keeper.go:
```diff
@@ -0,0 +1,82 @@
+	Granter   sdk.AccAddress `json:"granter"`
+}
+
+func (k Keeper) GetFeeAllowances(ctx sdk.Context, grantee sdk.AccAddress) []FeeAllowanceGrant {
```
The pattern we've been following in the SDK is to instead provide/expose iterators with a cb function param instead. We can keep this method, but we should also have an iterator method which this would simply use.

alexanderbez commented on x/fee-delegation/keeper.go:
```diff
@@ -0,0 +1,82 @@
+	if allowance == nil {
+		return false
+	}
+	allow, updated, delete := allowance.Accept(fee, ctx.BlockHeader())
```
ditto

fedekunze commented:
@aaronc are you planning to resume work here or can we close this PR?

ethanfrey commented:
@aaronc I am happy to work on addressing PR comments and updating to new module spec here.

However, can you add a high-level doc.go file explaining the use and main interfaces. I would much prefer to have a clear requirements than just assuming what is coded is correct.

I guess the "proper" way now is doing an ADR, but at least some clear docs is good.

aaronc commented:
> @aaronc I am happy to work on addressing PR comments and updating to new module spec here.
> 
> However, can you add a high-level doc.go file explaining the use and main interfaces. I would much prefer to have a clear requirements than just assuming what is coded is correct.
> 
> I guess the "proper" way now is doing an ADR, but at least some clear docs is good.

@ethanfrey I have added a doc.go and also better doc comments on `BasicFeeAllowance` and `PeriodicFeeAllowance`. Hopefully that is enough to make this clearer.

aaronc commented:
Replaced by #5207
